dali/kernels/imgproc/sampler.h

```diff
@@ -30,165 +30,183 @@ namespace kernels {
 
 struct BorderClamp {};
 
-template <DALIInterpType interp, typename In>
+template <DALIInterpType interp, int spatial_ndim, typename In>
 struct Sampler;
 
-template <DALIInterpType interp, typename T>
-DALI_HOST_DEV Sampler<interp, T> make_sampler(const Surface2D<T> &surface) {
-  return Sampler<interp, T>(surface);
+template <DALIInterpType interp, typename In>
+using Sampler2D = Sampler<interp, 2, In>;
+
+template <DALIInterpType interp, typename In>
+using Sampler3D = Sampler<interp, 3, In>;
+
+
+template <DALIInterpType interp, int ndim, typename T>
+DALI_HOST_DEV
+Sampler<interp, ndim, std::remove_const_t<T>> make_sampler(const Surface<ndim, T> &surface) {
+  return Sampler<interp, ndim, std::remove_const_t<T>>(surface);
 }
 
 template <typename T>
-DALI_HOST_DEV T GetBorderChannel(const T *values, int c) {
+DALI_HOST_DEV DALI_FORCEINLINE T GetBorderChannel(const T *values, int c) {
   return values[c];
 }
 
 template <typename T>
-DALI_HOST_DEV constexpr
+DALI_HOST_DEV DALI_FORCEINLINE constexpr
 enable_if_t<!std::is_pointer<T>::value, T> GetBorderChannel(const T &value, int) {
   return value;
 }
 
-template <typename In>
-struct Sampler<DALI_INTERP_NN, In> {
+/**
+ * @brief Checks if all coordinates in `coords` are positive and below the respective
+ *        coordinate in `limit`.
+ *
+ * @remarks `limit` is expected to have non-negative coordinates - otherwise, the result
+ *          is undefined.
+ */
+template <int n>
+DALI_HOST_DEV DALI_FORCEINLINE
+constexpr bool all_in_range(ivec<n> coords, ivec<n> limit) {
+  for (int i = 0; i < n; i++) {
+    if (static_cast<unsigned>(coords[i]) >= static_cast<unsigned>(limit[i]))
+      return false;
+  }
+  return true;
+}
+
+/**
+ * @brief Checks if all coordinates in `coords` are positive and below the respective
+ *        coordinate in `limit`.
+ *
+ * @remarks `limit` is expected to have non-negative coordinates - otherwise, the result
+ *          is undefined.
+ */
+DALI_HOST_DEV DALI_FORCEINLINE
+constexpr bool all_in_range(ivec2 coords, ivec2 limit) {
+  return static_cast<unsigned>(coords.x) < static_cast<unsigned>(limit.x) &&
+         static_cast<unsigned>(coords.y) < static_cast<unsigned>(limit.y);
+}
+
+/**
+ * @brief Checks if all coordinates in `coords` are positive and below the respective
+ *        coordinate in `limit`.
+ *
+ * @remarks `limit` is expected to have non-negative coordinates - otherwise, the result
+ *          is undefined.
+ */
+DALI_HOST_DEV DALI_FORCEINLINE
+constexpr bool all_in_range(ivec3 coords, ivec3 limit) {
+  return static_cast<unsigned>(coords.x) < static_cast<unsigned>(limit.x) &&
+         static_cast<unsigned>(coords.y) < static_cast<unsigned>(limit.y) &&
+         static_cast<unsigned>(coords.z) < static_cast<unsigned>(limit.z);
+}
+
+template <int _spatial_ndim, typename In>
```

Comment:
any objections with putting `spatial_ndim` last? The reason I bring that up is consistency with other Kernels where we usually have `OutType, InType, dims` in that order

Reply:
It's extending the Surface's parameters by adding interpolation type at the front - and I add this extra argument at the front to be consistent with make_sampler, which only needs one explicit template argument.